Add resize_image_aspect_with_pad method

danosipov opened the pull request:
This method adds the ability to resize the image to specified size with padding without distortions by preserving the aspect ratio.

Commit:
Document additional argument

martinwicke commented on tensorflow/python/ops/image_ops.py:
```diff
@@ -34,6 +34,7 @@
 @@resize_bilinear
 @@resize_nearest_neighbor
 @@resize_image_with_crop_or_pad
+@@resize_image_aspect_with_pad
```
Why is "aspect" in the name? I'd just call this `resize_image_with_pad`

danosipov replied:
I think the name should highlight the difference from the other `resize_*` methods, all of which distort the image if the new dimensions don't match the aspect ratio.

martinwicke replied:
`resize_image_with_crop_or_pad` doesn't distort either. I think calling it `resize_image_with_pad` would be better.

martinwicke commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -947,6 +947,119 @@ def resize_images(images,
+
+  Resizes an image to a target width and height by keeping
+  the aspect ratio the same without distortion and padding
+  it evenly with zeros.
```
Mention in more detail what happens: use resize to resize to the closes (what's that) size that can be achieved without distortion, then add padding.

martinwicke commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -947,6 +947,119 @@ def resize_images(images,
+    padded_height = height + (p_height * 2)
+    padded_width = width + (p_width * 2)
+
+    # Pad first, then resize to meet requested dimensions
```
I would have expected you pad later... is there a reason you pad first (which is more expensive, and will make padding bleed into the border of the image)?

danosipov replied:
Hey @martinwicke - thanks for the review! I'll address the comments. For this one I had trouble getting resize to work on arbitrary sizes without cropping. I'll try to recall the issue & add a comment explaining it better.

martinwicke commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -947,6 +947,119 @@ def resize_images(images,
+    assert_ops += _assert(
+        equal_(resized_height, target_height), ValueError,
+        'resized height is not correct.')
+    assert_ops += _assert(
```
These are debug asserts -- your method should be guaranteed to work, I don't think we'd want to keep these.

martinwicke commented:
(good for API review with name change)

Commit:
Merge pull request #18944 from tensorflow/martinwicke-patch-3

Fix docs rendering in placeholder docs page.

Commit:
Rename API method

Commit:
Remove assertions

Commit:
Update docstring

danosipov commented:
@martinwicke Updated, let me know if there are other changes you'd like to see.

Commit:
Fix bad merge

tensorflowbutler commented:
Nagging Assignee @martinwicke: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

Commit:
Resize first, pad second

Commit:
Remove unused function

martinwicke commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -969,6 +969,104 @@ def resize_images(images,
+    ratio = max_(f_width / f_target_width, f_height / f_target_height)
+    resized_height_float = f_height / ratio
+    resized_width_float = f_width / ratio
+    resized_height = math_ops.cast(math_ops.floor(p_height_float), dtype=dtypes.int32)
```
this won't work (p_*_float isn't defined)

danosipov replied:
I've had some issues with bazel recently, trying to fix it (and will fix this issue).

Is CI hooked up to github repo to test the branch? I see a `ubuntu` check pending, but doesn't look like it runs.

martinwicke replied:
It'll only run when I trigger it. I'll do that once there's a chance it'll pass.

martinwicke reviewed (changes requested):
Can you run your tests? There's at least one syntax error (undefined variable), which should be hit.

Commit:
Fix bugs

danosipov commented:
@martinwicke Sorry for the delay, the tests pass now.

I addressed all issues expect for https://github.com/tensorflow/tensorflow/pull/17814#discussion_r184870802 - can you elaborate? I checked `_CheckAtLeast3DImage` but its not clear to me how it can be used to replace assertions...

martinwicke commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -969,6 +969,110 @@ def resize_images(images,
+          assert_ops, target_height)
+    if _is_tensor(target_width):
+      target_width = control_flow_ops.with_dependencies(assert_ops,
+                                                        target_width)
```
Sorry, you can remove the "if _is_tensor", since `assert_ops` will not contain the checks for static target_* anyway.

So, just remove the two ifs above and you're ok.

Commit:
Remove unnecessary assertions

drpngx commented:
Please fix the lint errors:
```
FAIL: Found 8 non-whitelited pylint errors:
tensorflow/python/ops/image_ops_impl.py:1047: [C0330(bad-continuation), ] Wrong continued indentation (remove 7 spaces).
tensorflow/python/ops/image_ops_impl.py:1114: [C0301(line-too-long), ] Line too long (92/80)
tensorflow/python/ops/image_ops_impl.py:1115: [C0301(line-too-long), ] Line too long (90/80)
tensorflow/python/ops/image_ops_impl.py:1117: [C0301(line-too-long), ] Line too long (83/80)
tensorflow/python/ops/image_ops_test.py:2606: [C0330(bad-continuation), ] Wrong continued indentation (remove 6 spaces).
tensorflow/python/ops/image_ops_test.py:2617: [C0330(bad-continuation), ] Wrong continued indentation (remove 8 spaces).
tensorflow/python/ops/image_ops_test.py:2637: [C0330(bad-continuation), ] Wrong continued indentation (remove 6 spaces).
tensorflow/python/ops/image_ops_test.py:2653: [C0330(bad-continuation), ] Wrong continued indentation (remove 6 spaces).
```
and update the goldens:
```
    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```
After that it should be ready to go!

Commit:
Fix lint errors

danosipov commented:
@drpngx Lint should be good, and I ran the goldens steps

Commit:
Merge remote-tracking branch 'origin/master' into danosipov/feature/resize_image_aspect_with_pad

danosipov commented:
Hmm, going to try to rerun the api compatibility steps

danosipov commented:
Running update steps doesn't seem to change the repository - should it?

```
 ~/w/t/tensorflow   danosipov/feature/resize_image_aspect_with_pad …  git status                                                 Wed Jun  6 17:02:44 2018
On branch danosipov/feature/resize_image_aspect_with_pad
nothing added to commit but untracked files present (use "git add" to track)
 ~/w/t/tensorflow   danosipov/feature/resize_image_aspect_with_pad …  bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens True
sWARNING:tensorflow:VARIABLES collection name is deprecated, please use GLOBAL_VARIABLES instead; VARIABLES will be removed after 2017-03-02.
..
----------------------------------------------------------------------
Ran 3 tests in 0.279s

OK (skipped=1)
 ~/w/t/tensorflow   danosipov/feature/resize_image_aspect_with_pad …  git status                                        1601ms  Wed Jun  6 17:02:58 2018
On branch danosipov/feature/resize_image_aspect_with_pad
nothing added to commit but untracked files present (use "git add" to track)
```

martinwicke commented:
hmm... it should. Can you run this on Linux? I think it should now work for both Py2 and Py3, but to be sure, can you try with Py2?

Commit:
Merge remote-tracking branch 'origin/master' into danosipov/feature/resize_image_aspect_with_pad

danosipov commented:
Python 3 is a problem, but I get c++ compilation errors when I switch virtualenv to python 2:

```
ERROR: /Users/danosipov/workspace/tensorflow/tensorflow/tensorflow/python/BUILD:387:1: C++ compilation of rule '//tensorflow/python:py_util' failed (Exit 1): wrapped_clang failed: error executing command
  (cd /private/var/tmp/_bazel_danosipov/6485f105ce913a72c808d415e09f8e49/execroot/org_tensorflow && \
  exec env - \
    APPLE_SDK_PLATFORM='' \
    APPLE_SDK_VERSION_OVERRIDE='' \
    PATH=/Users/danosipov/workspace/tensorflow/tensorflow/tf/bin:/Users/danosipov/.rbenv/shims:/opt/chefdk/bin:/Users/danosipov/.chefdk/gem/ruby/2.3.0/bin:/opt/chefdk/embedded/bin:/usr/local/opt/coreutils/libexec/gnubin:/Users/danosipov/workspace/datadog/devtools/bin:/usr/local/opt/llvm/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/chefdk/gitbin \
    XCODE_VERSION_OVERRIDE=9.4.0 \
  external/local_config_cc/wrapped_clang '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections -g0 '-std=c++11' -g0 -iquote . -iquote bazel-out/host/genfiles -iquote external/com_google_absl -iquote bazel-out/host/genfiles/external/com_google_absl -iquote external/bazel_tools -iquote bazel-out/host/genfiles/external/bazel_tools -iquote external/nsync -iquote bazel-out/host/genfiles/external/nsync -iquote external/eigen_archive -iquote bazel-out/host/genfiles/external/eigen_archive -iquote external/local_config_sycl -iquote bazel-out/host/genfiles/external/local_config_sycl -iquote external/gif_archive -iquote bazel-out/host/genfiles/external/gif_archive -iquote external/jpeg -iquote bazel-out/host/genfiles/external/jpeg -iquote external/protobuf_archive -iquote bazel-out/host/genfiles/external/protobuf_archive -iquote external/com_googlesource_code_re2 -iquote bazel-out/host/genfiles/external/com_googlesource_code_re2 -iquote external/farmhash_archive -iquote bazel-out/host/genfiles/external/farmhash_archive -iquote external/fft2d -iquote bazel-out/host/genfiles/external/fft2d -iquote external/highwayhash -iquote bazel-out/host/genfiles/external/highwayhash -iquote external/png_archive -iquote bazel-out/host/genfiles/external/png_archive -iquote external/zlib_archive -iquote bazel-out/host/genfiles/external/zlib_archive -iquote external/local_config_python -iquote bazel-out/host/genfiles/external/local_config_python -isystem external/bazel_tools/tools/cpp/gcc3 -isystem external/nsync/public -isystem bazel-out/host/genfiles/external/nsync/public -isystem external/eigen_archive -isystem bazel-out/host/genfiles/external/eigen_archive -isystem external/gif_archive/lib -isystem bazel-out/host/genfiles/external/gif_archive/lib -isystem external/protobuf_archive/src -isystem bazel-out/host/genfiles/external/protobuf_archive/src -isystem external/farmhash_archive/src -isystem bazel-out/host/genfiles/external/farmhash_archive/src -isystem external/png_archive -isystem bazel-out/host/genfiles/external/png_archive -isystem external/zlib_archive -isystem bazel-out/host/genfiles/external/zlib_archive -isystem external/local_config_python/python_include -isystem bazel-out/host/genfiles/external/local_config_python/python_include -MD -MF bazel-out/host/bin/tensorflow/python/_objs/py_util/tensorflow/python/lib/core/py_util.d '-frandom-seed=bazel-out/host/bin/tensorflow/python/_objs/py_util/tensorflow/python/lib/core/py_util.o' -D__CLANG_SUPPORT_DYN_ANNOTATION__ -DEIGEN_MPL2_ONLY '-DEIGEN_MAX_ALIGN_BYTES=64' -DTENSORFLOW_USE_ABSL -DTF_USE_SNAPPY '-isysroot __BAZEL_XCODE_SDKROOT__' -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c tensorflow/python/lib/core/py_util.cc -o bazel-out/host/bin/tensorflow/python/_objs/py_util/tensorflow/python/lib/core/py_util.o)
In file included from tensorflow/python/lib/core/py_util.cc:20:
In file included from ./tensorflow/core/lib/core/errors.h:19:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:492:15: error: C++ requires a type specifier for all declarations
    char_type toupper(char_type __c) const
              ^
bazel-out/host/genfiles/external/local_config_python/python_include/pyport.h:731:29: note: expanded from macro 'toupper'
#define toupper(c) towupper(btowc(c))
                            ^
In file included from tensorflow/python/lib/core/py_util.cc:20:
In file included from ./tensorflow/core/lib/core/errors.h:19:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:498:48: error: too many arguments provided to function-like macro invocation
    const char_type* toupper(char_type* __low, const char_type* __high) const
                                               ^
bazel-out/host/genfiles/external/local_config_python/python_include/pyport.h:731:9: note: macro 'toupper' defined here
#define toupper(c) towupper(btowc(c))
        ^
In file included from tensorflow/python/lib/core/py_util.cc:20:
In file included from ./tensorflow/core/lib/core/errors.h:19:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:498:29: error: expected ';' at end of declaration list
    const char_type* toupper(char_type* __low, const char_type* __high) const
                            ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:510:48: error: too many arguments provided to function-like macro invocation
    const char_type* tolower(char_type* __low, const char_type* __high) const
....
```

I'll try to compile it on Linux next

Commit:
Merge remote-tracking branch 'origin/master' into danosipov/feature/resize_image_aspect_with_pad

Commit:
Update goldens

danosipov commented:
Updating from master finally resolved the issue, and I was able to update goldens:
https://github.com/tensorflow/tensorflow/pull/17814/commits/e2617ac25490b33c87b8e792eee0670b09a7305f

LMK if anything else is needed

martinwicke commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -969,6 +969,110 @@ def resize_images(images,
+  Resizes an image to a target width and height by keeping
+  the aspect ratio the same without distortion. If the target
+  dimensions don't match the image dimensions, the image
+  is padded with zeroes prior to resizing.
```
Can you update the docstring as well to reflect the computation order change?

danosipov replied:
Done

danosipov committed:
Update docstring for accuracy

danosipov commented:
🎆 Thanks for your patience @martinwicke !

martinwicke commented:
Thanks for your persistence!
